### PR TITLE
PLANNER-2391 CS-D: Don't go through constraint session

### PR DESCRIPTION
**Pull request by triceo**
For CS-D, the indirection of SDF -> `ConstraintSession` -> `KieSession` is needless waste of CPU cycles.

For CS-D, the chain becomes SDF -> `KieSession`.
For CS-B, the chain becomes SDF -> `BavetConstraintSession`.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSessionFactory.java [hunk 1]
```diff
@@ -28,6 +28,6 @@
      * @param workingSolution if null, uniform synthetic constraint weights will be applied
      * @return never null
      */
-    ConstraintSession<Solution_, Score_> buildSession(boolean constraintMatchEnabled, Solution_ workingSolution);
+    Object buildSession(boolean constraintMatchEnabled, Solution_ workingSolution);
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSessionFactory.java:31]: `KieSession`/`BavetConstraintSession` have nothing in common.
Handling of this happens in `*ConstraintStreamScoreDirectorFactory` where it becomes fully type-safe.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/BavetConstraintStreamScoreDirector.java [hunk 2]
```diff
@@ -37,12 +37,13 @@
  * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
  * @see ScoreDirector
  */
-public class ConstraintStreamScoreDirector<Solution_, Score_ extends Score<Score_>>
-        extends AbstractScoreDirector<Solution_, Score_, AbstractConstraintStreamScoreDirectorFactory<Solution_, Score_>> {
+public final class BavetConstraintStreamScoreDirector<Solution_, Score_ extends Score<Score_>>
+        extends AbstractScoreDirector<Solution_, Score_, BavetConstraintStreamScoreDirectorFactory<Solution_, Score_>> {
 
-    protected ConstraintSession<Solution_, Score_> session;
+    protected BavetConstraintSession<Solution_, Score_> session;
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/BavetConstraintStreamScoreDirector.java:43]: Can't we roll the BavetConstraintSession also into BavetConstraintStreamScoreDirector?

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/BavetConstraintStreamScoreDirector.java:43]: Ok, maybe for a later PR.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java [hunk 3]
```diff
@@ -65,21 +63,6 @@ protected AbstractConstraintStreamScoreDirectorFactory(SolutionDescriptor<Soluti
                 (ConstraintSessionFactory<Solution_, Score_>) constraintFactory.buildSessionFactory(constraints);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/stream/AbstractConstraintStreamScoreDirectorFactory.java:63]: This line should now happen in the drools/bavet subclasses. Those should also no longer doe the fancy supplier trick, just the validation of the constraints can go through a common method.

**Review by ge0ffrey — APPROVED** [review]
Approved if buildSessionFactory() doesn't return an Object

**Comment by triceo** [comment]
I agree with the comments, they will be addressed in a follow-up PR as agreed.

**Pull request merged**